### PR TITLE
Implementation of TokenEscrow contract (#1099)

### LINKED ISSUE #1099: Extend Escrows to use tokens
Currently, the [Escrow](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/release-v1.11.0/contracts/payment/Escrow.sol) contracts can only be used to transfer Ether. #1010 introduced a contract very similar to a [ConditionalEscrow](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/release-v1.11.0/contracts/payment/ConditionalEscrow.sol), with the added bonus of also allowing tokens to be transferred.

Do we want to extend the current `Escrow` so that it can also use tokens? The original intent behind it was to enable [async payments](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/release-v1.11.0/contracts/payment/PullPayment.sol), which is not required for an ERC20 token (since in both cases it'd boil down to a `transfer` call). It is when `ConditionalEscrow`s are considered that the benefits emerge: we could have [refund payments](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/release-v1.11.0/contracts/payment/RefundEscrow.sol) with tokens, the aforementioned timelock, etc.

If there's interest in this idea, I can try out some designs and come back with a more concrete proposal (in terms of implementation details).

### PR DESCRIPTION
**Pull request by tinchoabbate**
Fixes #1099

# 🚀 Description

Implementation of the `TokenEscrow` contract. Unlike the existing `Escrow` contract which handles ether, `TokenEscrow` can be used to put any ERC20 token in escrow, as long as it is previously approved by the holder of the tokens. Appart from that, the use case of `TokenEscrow` as a payment method should be pretty similar to `Escrow`.

- [x] 📘 I've reviewed the [OpenZeppelin Contributor Guidelines](../blob/master/CONTRIBUTING.md)
- [x] ✅ I've added tests where applicable to test my new functionality.
- [x] 📖 I've made sure that my contracts are well-documented.
- [x] 🎨 I've run the JS/Solidity linters and fixed any issues (`npm run lint:all:fix`).

### CODE REVIEW

File: test/payment/TokenEscrow.test.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            await this.tokenEscrow.deposit(payer, payee, amount, { from: owner });
+            (await this.tokenEscrow.depositsOf(payee)).should.be.bignumber.equal(amount);
+          }
+
+          (await this.standardToken.balanceOf(this.tokenEscrow.address)).should.be.bignumber.equal(amount * 2);
+        });
+      });
+
+      describe('withdrawals', function () {
+        it('can withdraw payments', async function () {
```
Comment (nventuro) [inline test/payment/TokenEscrow.test.js]: These tests may benefit from a 'with deposit' context.

File: contracts/payment/TokenEscrow.sol [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+contract TokenEscrow is Ownable {
+  using SafeMath for uint256;
+  using SafeERC20 for ERC20;
+
+  event Deposited(address indexed payee, uint256 tokenAmount);
+  event Withdrawn(address indexed payee, uint256 tokenAmount);
+
+  mapping(address => uint256) private deposits;
+
+  ERC20 token;
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/payment/TokenEscrow.sol]: Also, we should add a test that reads from this variable and compares the addresses.

File: contracts/payment/TokenEscrow.sol [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * to the TokenEscrow's deposit and withdraw.
+ * Moreover, the TokenEscrow should also be allowed to transfer
+ * tokens from the payer to itself.
+ */
+contract TokenEscrow is Ownable {
+  using SafeMath for uint256;
+  using SafeERC20 for ERC20;
+
+  event Deposited(address indexed payee, uint256 tokenAmount);
+  event Withdrawn(address indexed payee, uint256 tokenAmount);
```
[the remark this answers is not in the document]
Comment (frangio) [reply contracts/payment/TokenEscrow.sol]: Hm, I'd keep them.

File: contracts/payment/TokenEscrow.sol [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  /**
+  * @dev Puts in escrow a certain amount of tokens as credit to be withdrawn.
+  * @param _payer Address holding tokens which will be put in escrow.
+  * @param _payee The destination address of the tokens.
+  * @param _amount The amount of tokens to deposit in escrow.
+  */
+  function deposit(
+    address _payer,
```
[the remark this answers is not in the document]
Comment (tinchoabbate) [reply contracts/payment/TokenEscrow.sol]: Yeap, you've got a point there, my bad. Thanks for making it clearer :+1:

**Comment by nventuro** [comment]
Since it looks like we may break some things along the way, I'm creating a new `escrow-exploration` branch, and changing to PR to merge there.

File: test/payment/TokenEscrow.test.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+const expectEvent = require('../helpers/expectEvent');
+
+const BigNumber = web3.BigNumber;
+
+require('chai')
+  .use(require('chai-bignumber')(BigNumber))
+  .should();
+
+const TokenEscrow = artifacts.require('TokenEscrow');
+const token = artifacts.require('StandardTokenMock');
```
[the remark this answers is not in the document]
Comment (tinchoabbate) [reply test/payment/TokenEscrow.test.js]: Nice catch! Hadn't noticed it :sweat_smile:  The consequences of a batch find & replace.

**Pull request merged**